coral-incremental/src/main/java/com/linkedin/coral/incremental/RelNodeCostEstimator.java

```diff
@@ -0,0 +1,285 @@
+          Double distinctCount = distinctEntry.getValue().getAsDouble();
+
+          tableStatistic.distinctCountByRow.put(columnName, distinctCount);
+        }
+        costStatistic.put(tableName, tableStatistic);
+
+      }
+    } catch (IOException e) {
+      throw new IOException("Failed to load statistics from the configuration file: " + configPath, e);
+    }
+
+  }
+
+  /**
+   * Returns the total cost of executing a relational operation.
+   *
+   * <p>This method computes the cost of executing a relational operation based on the input
+   * relational expression. The cost is calculated as the sum of the execution cost and the I/O cost.
+   * We assume that I/O only occurs at the root of the query plan (Project) where we write the output to disk.
+   * So the cost is the sum of the shuffle cost of all children RelNodes and IOCostValue * row count of the root Project RelNode.
+   *
+   * @param rel the input relational expression
+   * @return the total cost of executing the relational operation
+   */
+  public Double getCost(RelNode rel) {
+    CostInfo executionCostInfo = getExecutionCost(rel);
+    Double IOCost = executionCostInfo.rowCount * IOCostValue;
+    return executionCostInfo.shuffleCost * shuffleCostValue + IOCost;
+  }
+
+  public CostInfo getExecutionCost(RelNode rel) {
+    if (rel instanceof TableScan) {
+      return getExecutionCostTableScan((TableScan) rel);
+    } else if (rel instanceof LogicalJoin) {
+      return getExecutionCostJoin((LogicalJoin) rel);
+    } else if (rel instanceof LogicalUnion) {
+      return getExecutionCostUnion((LogicalUnion) rel);
+    } else if (rel instanceof LogicalProject) {
+      return getExecutionCostProject((LogicalProject) rel);
+    }
+    throw new IllegalArgumentException("Unsupported relational operation: " + rel.getClass().getSimpleName());
+  }
+
+  private CostInfo getExecutionCostTableScan(TableScan scan) {
+    RelOptTable originalTable = scan.getTable();
+    String tableName = getTableName(originalTable);
+    try {
+      TableStatistic tableStat = costStatistic.get(tableName);
+      Double rowCount = tableStat.rowCount;
+      return new CostInfo(rowCount, rowCount);
+    } catch (NullPointerException e) {
+      throw new IllegalArgumentException("Table statistics not found for table: " + tableName);
+    }
+  }
+
+  private String getTableName(RelOptTable table) {
+    return String.join(".", table.getQualifiedName());
+  }
+
+  private CostInfo getExecutionCostJoin(LogicalJoin join) {
+    RelNode left = join.getLeft();
+    RelNode right = join.getRight();
+    CostInfo leftCost = getExecutionCost(left);
+    CostInfo rightCost = getExecutionCost(right);
+    Double joinSize = estimateJoinSize(join, leftCost.rowCount, rightCost.rowCount);
+    // The shuffle cost of a join is the maximum shuffle cost of its children because
+    // in modern distributed systems, the shuffle cost is dominated by the largest shuffle.
```

Comment:
Terminology is not clear here. What is the shuffle cost of children? We might need to discuss this over a doc.